### PR TITLE
Dlo 20rework

### PR DESCRIPTION
**Pull request by Willning**
### **Related Issue/Keyword:**
 Mapping Screen prototype completion. 
### **Description:**
Yeah, this the biggie. Project mapping and saving working. Can now create a mapping with a medium restricting modes for that plan. Also loading from projects and recent projects work. 

### **Testing:**
Make a plan and save, should load. 

**Important note** 
Any previous data in the DB will be deprecated and won't load properly due to format changes. 

**Steps for manual testing:**

1. Make a plan and add mappings etc. 
2. Save
3. Load plan and everything should be there. 

### **Checklist:**

- [x] Latest master merged/rebased into your feature branch 
- [x] Meets the projects coding conventions
- [x] No out of scope changes
- [x] #Mentioned other relevant issues
- [x] No failure when running the application (obviously lol)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
     "bootstrap": "^4.3.1",
     "firebase": "^6.4.0",
     "history": "^4.9.0",
+    "orgchart": "^2.1.5",
```
Comment (hybrio) [inline package.json]: i don't think orgchart is used, plz remove

File: src/Firestore.js [hunk 2]
```diff
@@ -44,11 +44,13 @@ class Firestore {
...
 
     static getProjectById(userID, projectID) {
+        //Can't seem to do doc on the returned promise
         return this.getAllProjectsByUser(userID).doc(projectID);
     };
 
     static getAllIdeasByProject(userID, projectID) {
-        return this.getProjectById(userID, projectID).collection("Ideas");
+        //return this.getProjectById(userID, projectID).collection("ideas");
+        return db.collection("users").doc(userID).collection("projects").doc(projectID).collection("ideas");
```
Comment (hybrio) [inline src/Firestore.js]: why doesn't this use this.getProjectById(userID, projectID).collection("ideas") instead?

File: src/Firestore.js [hunk 3]
```diff
@@ -60,13 +62,31 @@ class Firestore {
...
         var ideaCollection = this.getAllIdeasByProject(userID, projectID);
         ideas.forEach(function(idea) {
             this.saveToDBWithDocID(ideaCollection, idea.id, idea);
         });
+        
     };
 
+    //save a singular idea to the project.
+    static saveIdeaToProject(userID, projectID, ideaID, idea) {
+        var ideaRef = db.collection("users").doc(userID).collection("projects").doc(projectID).collection("ideas").doc(ideaID);
```
Comment (hybrio) [inline src/Firestore.js]: again this should probably get the reference from this.getProjectById(userID, projectID).collection("ideas")

File: src/Firestore.js [hunk 4]
```diff
@@ -60,13 +62,31 @@ class Firestore {
...
+        return ideaRef.set({
+                title: idea.title,
+                mode: idea.mode,
+                notes: idea.notes,
+                parentID: idea.parentID,
+            })
+    }
+
+    static deleteIdeafromProject(userID, projectID, ideaID){
+        var ideaRef = db.collection("users").doc(userID).collection("projects").doc(projectID).collection("ideas").doc(ideaID).delete();
```
Comment (hybrio) [inline src/Firestore.js]: same issue with this.getProjectById(userID, projectID).collection("ideas") as previous comments

File: src/Firestore.js [hunk 5]
```diff
@@ -79,6 +99,11 @@ class Firestore {
         return db.collection("users").doc(userID).collection("projects").add(projectData);
     }
 
+    static editProjectFields(userID, projectID, data) {
+        var ideaRef = db.collection('users').doc(userID).collection("projects").doc(projectID);
```
Comment (hybrio) [inline src/Firestore.js]: again should probably get project reference from this.getProjectById(userID, projectID)

File: src/IdeaEditPopup.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    handleNotesChange = (event)=> {
+        this.setState({notes: event.target.value});
+    }
+    
+    handleModeChange = (mode) => {
+        this.setState({mode:mode}); 
+    }
+
+    render(){
+        //@@TODO Layout the dropdown to look nicer.
```
Comment (hybrio) [inline src/IdeaEditPopup.js]: isn't this already done? (so maybe remove code comment)

File: package.json (hunk 1 above)
Comment (Willning) [reply package.json]: yes ok

File: src/Firestore.js (hunk 2 above)
Comment (Willning) [reply src/Firestore.js]: Can't seem to append more specifiers once the query has been returned by this.getProjectId.  (i.e. doing .doc() or .collection() on the returned query breaks.

Comment (Willning) [reply src/Firestore.js]: nope, seem to have fixed it. converting now

**Pull request merged**